File: kotlin/rotational-cipher/src/main/kotlin/RotationalCipher.kt
class RotationalCipher(rotations: Int) {
    private val map = ('a'..'z').shiftMap(rotations) +
                      ('A'..'Z').shiftMap(rotations)

    fun encode(text: String) = text.map { map.getOrDefault(it, it) }.joinToString("")

    private fun CharRange.shiftMap(n: Int) = zip(shift(n)).toMap()
    private fun CharRange.shift(n: Int) = drop(n) + take(n)
}

File: elixir/rotational-cipher/lib/rotational_cipher.ex
defmodule RotationalCipher do
  @lower_alphabet Enum.to_list(?a..?z)
  @upper_alphabet Enum.to_list(?A..?Z)

  def rotate(text, shift) do
    letter_map = rotate_map(shift)

    text
    |> String.to_charlist()
    |> Enum.map(&Map.get(letter_map, &1, &1))
    |> List.to_string()
  end

  defp rotate_map(shift) do
    Map.merge(shift_map(@lower_alphabet, shift), shift_map(@upper_alphabet, shift))
  end

  defp shift_map(enum, shift) do
    Enum.zip(enum, shift_by(enum, shift)) |> Map.new()
  end

  defp shift_by(enum, by) do
    Enum.drop(enum, by) ++ Enum.take(enum, by)
  end
end
